### PR TITLE
MCS Login and Authentication

### PR DESCRIPTION
**Pull request by dvaldivia**
Adds the authorization requirement to all API calls except for `/api/v1/login` (GET and POST)

Upon login-in at `api/v1/login` with a sample payload
```
{
"accessKey":"minioadmin",
"secretKey":"minioadmin"
}
```

One must present the returned session on the `Authorization: Bearer token` header

### CODE REVIEW

File: mcs/restapi/users_login.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	})
+
+}
+
+var ErrInvalidCredentials = errors.New("invalid credentials")
+
+// login performs a check of credentials against MinIO
+func login(accessKey, secretKey *string) (*string, error) {
+	// Probe the credentials
+	cfg, pErr := mcCmd.BuildS3Config(getMinIOServer(), *accessKey, *secretKey, "", "auto")
```
Comment (cesnietor) [inline mcs/restapi/users_login.go]: mcCmd can be used as a parameter the same way we do it for  all other apis with MinioClient .That way we can add some tests for this.

**Review by cesnietor — CHANGES_REQUESTED** [review]

**Review by Alevsk — CHANGES_REQUESTED** [review]
Reviewing

File: mcs/swagger.yml [hunk 2]
```diff
@@ -8,6 +8,14 @@ produces:
...
 schemes:
   - http
+securityDefinitions:
+  key:
+    type: oauth2
+    flow: accessCode
+    authorizationUrl: http://min.io
+    tokenUrl: http://min.io
+security:
+  - key: []
```
Comment (cesnietor) [inline mcs/swagger.yml]: this is not being used

Comment (cesnietor) [inline mcs/swagger.yml]: can you please add comments `#...` next to this definitions to add more doc of what they do

File: mcs/swagger.yml [hunk 3]
```diff
@@ -753,4 +798,29 @@ definitions:
...
+      accessKey:
+        type: string
+      secretKey:
+        type: string
+  loginResponse:
+    type: object
+    properties:
+      sessionId:
+        type: string
+  principal:
```
Comment (cesnietor) [inline mcs/swagger.yml:847]: can you add a comment of what this is for?

File: mcs/restapi/users_login.go (hunk 1 above)
Comment (dvaldivia) [reply mcs/restapi/users_login.go]: I don't think we can mock imports as they will be linked at compile time, mcCmd is not an object per se

File: mcs/swagger.yml (hunk 2 above)
Comment (dvaldivia) [reply mcs/swagger.yml]: sure

File: mcs/swagger.yml (hunk 3 above)
Comment (dvaldivia) [reply mcs/swagger.yml:847]: done

File: mcs/restapi/sessions/sessions.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	return sessionId, nil
+}
+
+func (s *Singleton) ValidSession(sessionId string) bool {
+	if _, ok := s.sessions[sessionId]; ok == false {
+		return true
+	}
+	return false
+}
+
```
Comment (cesnietor) [inline mcs/restapi/sessions/sessions.go:62]: can you please add tests for all this functions?

**Review by Alevsk — DISMISSED** [review]

File: mcs/restapi/sessions/sessions.go (hunk 4 above)
Comment (dvaldivia) [reply mcs/restapi/sessions/sessions.go:62]: done

**Review by Alevsk — DISMISSED** [review]

**Review by cesnietor — COMMENTED** [review]
lint errors
```
golangci-lint run --timeout=5m --config ../.golangci.yml
restapi/sessions/sessions.go:49:2: var `sessionId` should be `sessionID` (golint)
	sessionId := RandomCharString(64)
	^
restapi/sessions/sessions.go:60:34: method parameter `sessionId` should be `sessionID` (golint)
func (s *Singleton) ValidSession(sessionId string) bool {
                                 ^
restapi/sessions/sessions_test.go:32:2: var `sessionId` should be `sessionID` (golint)
	sessionId, err := GetInstance().NewSession(&cfg)
	^
restapi/sessions/sessions_test.go:42:2: var `sessionId` should be `sessionID` (golint)
	sessionId, _ := GetInstance().NewSession(&cfg)
	^
restapi/user_login.go:103:55: getLoginDetailsResponse - result 1 (error) is always nil (unparam)
func getLoginDetailsResponse() (*models.LoginDetails, error) {
                                                      ^
restapi/sessions/sessions.go:51:37: S1002: should omit comparison to bool constant, can be simplified to `!ok` (gosimple)
	if _, ok := s.sessions[sessionId]; ok == false {
	                                   ^
restapi/sessions/sessions.go:61:37: S1002: should omit comparison to bool constant, can be simplified to `ok` (gosimple)
	if _, ok := s.sessions[sessionId]; ok == true {
```

**Comment by dvaldivia** [comment]
@cesnietor done

File: mcs/restapi/user_login_test.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+var mcBuildS3ConfigMock func(url, accessKey, secretKey, api, lookup string) (*mcCmd.Config, *probe.Error)
+
+type mcCmdMock struct{}
+
+func (mc mcCmdMock) BuildS3Config(url, accessKey, secretKey, api, lookup string) (*mcCmd.Config, *probe.Error) {
+	return mcBuildS3ConfigMock(url, accessKey, secretKey, api, lookup)
+}
+
+func TestLogin(t *testing.T) {
```
Comment (cesnietor) [inline mcs/restapi/user_login_test.go:37]: Can you add Some comments on each test similar to how we are doing for all other functions, else it is not clear what is expected or what we are trying to test. (test number, what we are testing, and the name of the function)

File: mcs/restapi/user_login_test.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	access := "ABCDEFHIJK"
+	secret := "ABCDEFHIJKABCDEFHIJK"
+	mcBuildS3ConfigMock = func(url, accessKey, secretKey, api, lookup string) (config *mcCmd.Config, p *probe.Error) {
+		return &mcCmd.Config{}, nil
+	}
+
+	sessionID, err := login(mcx, &access, &secret)
+	assert.NotEmpty(sessionID, "Session ID was returned empty")
+	assert.Nil(err, "error creating a session")
+}
```
Comment (cesnietor) [inline mcs/restapi/user_login_test.go:62]: also we need to test the `error ` scenarios, you can use coverage to see that the error cases are not being covered.

File: mcs/restapi/sessions/sessions_test.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+package sessions
+
+import (
+	"testing"
+
+	mcCmd "github.com/minio/mc/cmd"
+
+	"github.com/stretchr/testify/assert"
+)
+
```
Comment (cesnietor) [inline mcs/restapi/sessions/sessions_test.go:26]: Can you add Some comments on each test similar to how we are doing for all other functions, else it is not clear what is expected or what we are trying to test. (test number, what we are testing, and the name of the function)

File: mcs/restapi/sessions/sessions_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+func TestNewSession(t *testing.T) {
+	assert := assert.New(t)
+	// We will write a test against play
+	// Probe the credentials
+	cfg := mcCmd.Config{}
+	sessionID, err := GetInstance().NewSession(&cfg)
+	assert.NotEmpty(sessionID, "Session ID was returned empty")
+	assert.Nil(err, "error creating a session")
```
Comment (cesnietor) [inline mcs/restapi/sessions/sessions_test.go]: also we need to test the `error ` scenarios, you can use coverage to see that the error cases are not being covered.
```
go test ./restapi/sessions -coverprofile=coverage.out && go tool cover -html=coverage.out && open coverage.html
```

**Review by cesnietor — CHANGES_REQUESTED** [review]

**Review by bexsoft — DISMISSED** [review]
Test approve

File: mcs/restapi/user_login_test.go (hunk 6 above)
Comment (dvaldivia) [reply mcs/restapi/user_login_test.go:62]: The error is returned by the server (invalid, or not initialized, etc), so it'd be easier to test for error presence

File: mcs/go.mod [hunk 9]
```diff
@@ -18,3 +18,5 @@ require (
 	github.com/stretchr/testify v1.5.1
 	golang.org/x/net v0.0.0-20200301022130-244492dfa37a
 )
+
+replace github.com/minio/mc => github.com/dvaldivia/mc v0.0.0-20200330203654-e8aaf0b56ebd
```
Comment (cesnietor) [inline mcs/go.mod]: can we remove this :)

Comment (dvaldivia) [reply mcs/go.mod]: done

**Review by Alevsk — APPROVED** [review]

**Review by cesnietor — APPROVED** [review]

**Pull request merged**